consensus/src/chained_bft/block_storage/sync_manager.rs

```diff
@@ -187,6 +210,113 @@ impl<T: Payload> BlockStore<T> {
         self.rebuild(root, blocks, quorum_certs).await;
         Ok(())
     }
+
+    /// Check the highest ledger info sent by peer to see if we're behind and start
+    /// state synchronization if the committed block doesn't exist in the tree (the whole tree is
+    /// outdated).
+    async fn process_highest_ledger_info(
+        &self,
+        highest_ledger_info: QuorumCert,
+        retriever: &mut BlockRetriever,
+    ) -> failure::Result<()> {
+        let committed_block_id = highest_ledger_info
+            .committed_block_id()
+            .ok_or_else(|| format_err!("highest ledger info has no committed block"))?;
+        // Unfortunately we're leaking some information about the commit rule here:
+        // qc.parent_block() must be a child of the commit.
+        // Hence, commit_round is qc.parent_block() - 1.
+        let commit_round = if highest_ledger_info.parent_block().round() > 0 {
+            highest_ledger_info.parent_block().round() - 1
+        } else {
+            0
+        };
+        if !self.is_tree_outdated(committed_block_id, commit_round) {
+            return Ok(());
+        }
+        // TODO: store the pending LedgerInfo for the duration of the state synchronization
+        let mut attempt = 0_u32;
+        loop {
+            debug!(
+                "State sync attempt = {} to round {} from {}",
+                attempt,
+                commit_round,
+                self.root()
+            );
+            match self
+                .state_sync_attempt(highest_ledger_info.clone(), retriever)
+                .await
+            {
+                Err(e) => {
+                    error!(
+                        "Failed state sync attempt {} to round {}: {}",
+                        attempt, commit_round, e
+                    );
+                    attempt += 1;
+                }
+                Ok(_) => {
+                    debug!("State sync success. New root: {}", self.root());
+                    return Ok(());
+                }
+            }
+        }
+    }
+
+    async fn state_sync_attempt(
+        &self,
+        highest_ledger_info: QuorumCert,
+        retriever: &mut BlockRetriever,
+    ) -> failure::Result<()> {
+        // State computer synchronization must eventually succeed (an infinite retry):
+        // it returns the LedgerInfo corresponding to the latest committed state in accumulator.
+        let pre_sync_instance = Instant::now();
+        let new_root_li = self
+            .state_computer
+            .state_sync_or_bail(highest_ledger_info.ledger_info().clone());
```

Comment:
I realize this doesn't work well cross epoch with the current per-epoch structure.

With current design within a single EventProcessor we only care about one epoch, we only process messages from the same epoch including SyncInfo and will jump over at the epoch boundary if commit the end-epoch block and restart everything. 
If we receive future epoch message, we'll first request an epoch proof to jump to the genesis of that epoch and start new EventProcessor, so we'll never need to consider stuff from different epoch.

Although if state sync could return anything in the future, we'll need to handle epoch change here and https://github.com/libra/libra/blob/master/consensus/src/chained_bft/epoch_manager.rs#L115 (well, every call to state_sync)

We can brainstorm more about this tomorrow